packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -1435,11 +1456,60 @@ export function simplify<T>(
 					})),
 				],
 			}
+		case 'list':
+			if (!context.typeMapping) return typeDef
+			return {
+				...typeDef,
+				item: {
+					kind: 'mapped',
+					child: typeDef.item,
+					mapping: context.typeMapping,
+				},
+			}
+		case 'tuple':
+			if (!context.typeMapping) return typeDef
+			return {
+				...typeDef,
+				items: typeDef.items.map(item => ({
+					kind: 'mapped',
+					child: item,
+					mapping: context.typeMapping!,
+				})),
+			}
 		case 'enum':
 			return { ...typeDef, enumKind: typeDef.enumKind ?? 'int' }
-		case 'concrete': // TODO
-		case 'template': // TODO
-			return { kind: 'union', members: [] }
+		case 'concrete':
+			const simplifiedArgs = typeDef.typeArgs
+				.map(arg => simplify(arg, options, context))
```

Comment:
Ah, so in the following example? I think it makes sense that `[[id]]` refers to the field in `Bar` and not in `Foo`.
```rs
type Foo<T> = struct {
   id: string,
   components: T,
}

struct Bar {
   id: string,
   ...Foo<minecraft:data_component[[id]]>,
}
```
Simplifying the type here early is also the reason that the nested example from [07ae038](https://github.com/SpyglassMC/Spyglass/pull/1247/commits/07ae0387b3ea2921f86530b56a2f805dfdc81c11) works.

Reply:
Yep, exactly. Both make sense in their own way.

Like you have it now makes sense because this is the location of the dispatcher, so it makes some sense that the index should be from there.

The other one makes sense if the mental model is to replace all `T` with the exact type that the concrete type specifies.

Reply:
The test wouldn't work anymore because the `<[A]>` information would be lost when it enters the inner template. Only the most inner template mapping is kept. So if it tried to simplify `[A]` later, it wouldn't find the `A` reference.